lib/ts/recipe/multifactorauth/recipe.ts

```diff
@@ -0,0 +1,508 @@
+            same email as that of the second factor. For example, if first factor is google login with e1 
+            and second factor is email password with e2, we allow linking them as long as there is no other 
+            user with email e2.
+
+            We rejected this idea cause if auto account linking is switched off, then someone else can sign up 
+            with google using e2. This is OK as it would not link (since account linking is switched off). 
+            But, then if account linking is switched on, then the google sign in (and not sign up) with e2 
+            would actually cause it to be linked with the e1 account.
+        */
+
+        // we allow setup of unverified account info only if the session user has the same account info
+        // and it is verified
+
+        if (accountInfo.email !== undefined) {
+            let foundVerifiedEmail = false;
+            for (const lM of sessionUser?.loginMethods) {
+                if (lM.email === accountInfo.email && lM.verified) {
+                    foundVerifiedEmail = true;
+                    break;
+                }
+            }
+            if (!foundVerifiedEmail) {
+                return {
+                    status: "VALIDATION_ERROR",
+                    reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_010)",
+                };
+            }
+        }
+
+        if (accountInfo.phoneNumber !== undefined) {
+            let foundVerifiedPhoneNumber = false;
+            for (const lM of sessionUser?.loginMethods) {
+                if (lM.phoneNumber === accountInfo.phoneNumber) {
+                    foundVerifiedPhoneNumber = true;
+                    break;
+                }
+            }
+            if (!foundVerifiedPhoneNumber) {
+                throw new Error("should never happen"); // phone number only comes from passwordless and is a verified factor always
+            }
+        }
+
+        return {
+            status: "OK",
+        };
+    };
+
+    checkIfFactorUserCanBeLinkedWithSessionUser = async (
+        tenantId: string,
+        sessionUser: User,
+        accountInfo: { email?: string; phoneNumber?: string },
+        userContext: UserContext
+    ): Promise<{ status: "OK" | "RECURSE_FOR_RACE" } | { status: "VALIDATION_ERROR"; reason: string }> => {
+        if (!sessionUser.isPrimaryUser) {
+            const canCreatePrimary = await AccountLinkingRecipe.getInstance().recipeInterfaceImpl.canCreatePrimaryUser({
+                recipeUserId: sessionUser.loginMethods[0].recipeUserId,
+                userContext,
+            });
+
+            if (canCreatePrimary.status === "RECIPE_USER_ID_ALREADY_LINKED_WITH_PRIMARY_USER_ID_ERROR") {
+                // Race condition since we just checked that it was not a primary user
+                this.querier.invalidateCoreCallCache(userContext);
```

Comment:
why do we have to call this here explicitly? This is not good.. cause it's not at all obvious that we need to call this. If this is necessary to call here, have you missed calling it somewhere?

Reply:
not required, confirmed that there would be a non GET request that would have cleared the cache already